### PR TITLE
Post Featured Image block: Enable gradient overlay

### LINKED ISSUE #43676: Post Featured Image block: Enable gradient overlay
## What problem does this address?

While creating a design for Twenty Twenty-Three, I noticed it's not possible to add a gradient overlay to the featured images of posts, as the Post Featured Image block is used, where this feature is not available. It would be possible with using a Cover block, which can also display the featured image; however it's not intended to change the markup of Twenty Twenty-Three.

## What is your proposed solution?

To enhance the Post Featured Image block, the overlay feature from the Cover block should be made available. This would enable more design possibilities using this block, without having to rely on the Cover block instead.

### PR DESCRIPTION
**Pull request by RahiDroid**
## What?

- Adds the `Overlay` settings (color & opacity) to the Post featured block.

## Why?
- Checkout: https://github.com/WordPress/gutenberg/issues/43676 & https://github.com/WordPress/twentytwentythree/issues/30

## Testing scenarios covered
- Default gradients (Both inside site editor & the post editor)
- Custom gradients (Both inside site editor & the post editor)
- Default solid overlay color (Both inside site editor & the post editor)
- Custom solid overlay color (Both inside site editor & the post editor)
- Used inside the Query loop
- Didn't have a featured image on a post. (Placeholder will show the overlay as well)
- Above all at custom opacity levels and at 0.
- Above customizations with the Duotone applied. (Behaves same as the Cover block, i.e. applies an overlay on top of duotoned image).
- Overlay with the border settings applied.

## Testing Instructions

1. Open the post editor or the site editor.
1. Insert the `Post Featured Block`.
1. You should see the `Overlay` controls in the inspector controls of the block.
1. Applying the overlay should reflect both in the editor and on the front end.

## Screenshots or screencast 
Sep-03-2022 00-15-46

## Important notes
- When combined with the Duotone, it behaves similarly to the Cover block.
- The Cover block doesn't show the overlay on the placeholder inside the site editor. Either we add the overlay on the placeholder for the cover block, too, or we remove it from Post featured block to ensure consistent UX.

### CODE REVIEW

File: packages/block-library/src/post-featured-image/editor.scss
Comment (RahiDroid) [inline packages/block-library/src/post-featured-image/editor.scss:6]: This creates accessibility issue as the upload button could blend in with the overlay/background. However, having the upload button blurred background causes the overlay to look very different than how it actually would look like.

Need suggestions/opinions on this.

**Comment by getdave** [comment]
@mikachan as this is a Theme issue for TT3 would you be able to help out with a review? I'm also conscious that the Cover block now supports the featured image so it might be possible to achieve the same effect that way.

**Comment by RahiDroid** [comment]
> I'm also conscious that the Cover block now supports the featured image so it might be possible to achieve the same effect that way.

Just a thing to note, the Cover block requires a fixed height and can not have a dynamic height based on the featured image's original height. Using the Cover block in place of the Post Featured Image block could render it cropped. 

And by design, we can't assign dynamic height to the Cover block unless we do it conditionally when it has no inner blocks.

**Comment by critterverse** [comment]
Cool! Here's what I'm seeing:

https://user-images.githubusercontent.com/77359364/189169080-433b7909-1385-417b-8e29-8d69de9ed9d9.mp4

This is working really well IMO. I was worried that the flow might be awkward without the ability to define the dimratio on the overlay in theme.json, but this PR gets around that nicely by setting it to 0 to start. Nice work :)

**Review by mikachan — COMMENTED** [review]
Thank you for working on this! I've tested this out over various scenarios and it's working really well for me across the post editor, site editor, and front end.

> I'm also conscious that the Cover block now supports the featured image so it might be possible to achieve the same effect that way.

We're currently including 3 alternative template options in TT3: with no image, with a featured image, and with a cover block. This PR could enable us to reduce that back down to one template. There's some more info over in [this discussion](https://github.com/WordPress/twentytwentythree/issues/133).

>When combined with the Duotone, it behaves similarly to the Cover block.

This is the only thing that gives me pause, in that the cover block and the post featured image block are now very similar. I don't think it's a blocker at all, but perhaps a discussion for the future.

>The Cover block doesn't show the overlay on the placeholder inside the site editor. Either we add the overlay on the placeholder for the cover block, too, or we remove it from Post featured block to ensure consistent UX.

I'd personally like to see the cover block show the overlay in the editor, as that seems like an improved UX, but I think this could be a follow-up PR so as not to block this one.

**Comment by scruffian** [comment]
It looks like there's some code here that's also in the cover block. It might be worth considering if we can make this into a general solution that all blocks could use, rather than a specific per-block implementation.

**Comment by RahiDroid** [comment]
> It looks like there's some code here that's also in the cover block. It might be worth considering if we can make this into a general solution that all blocks could use, rather than a specific per-block implementation.

I agree with you  @scruffian, and I was wondering what that place would be though. And if we could do that extraction in a separate PR where we could also cover some refactoring to the Cover block, both in terms of code and feature (Placeholder overlay color). If we could ship this PR for the 14.1 RC1 release, it would be great, as this one could help us with some style variations on the TT3 theme?

**Comment by RahiDroid** [comment]
Hi @mikachan 👋 

Hope you had a good weekend, any hints on what the next steps would be for this one?

**Comment by mikachan** [comment]
It would be great if we could consolidate the functionality that overlaps with the cover block, but I agree this could happen in a follow-up PR along with any refactoring. We should also revisit the accessibility issue you mentioned [here](https://github.com/WordPress/gutenberg/pull/43838#discussion_r962382389). We should create an issue for this refactoring work so it's easy to track. 

Otherwise, I think this is good to bring in now! 🎉

**Review by mikachan — APPROVED** [review]

**Comment by RahiDroid** [comment]
Hi @mikachan, thank you for guiding me through this. I have created an issue for the accessibility issue https://github.com/WordPress/gutenberg/issues/44115. I will create an issue for the Cover block's refactoring once this PR gets merged as before we work on that, we could get some inputs on this accessibility issue as well.

Please let me know what would be the next step on this. Do we need any more approvals on this? Do you think we can land this in 14.1.0 RC? 🙈

File: packages/block-library/src/post-featured-image/editor.scss
Comment (RahiDroid) [reply packages/block-library/src/post-featured-image/editor.scss:6]: Have created separate issue https://github.com/WordPress/gutenberg/issues/44115 for tracking this.

**Comment by mikachan** [comment]
Thanks for creating that issue, and your plan for the refactoring PR sounds great! Please feel free to tag me, and I can help find other appropriate folks for review/input as well.

@c4rl0sbr4v0 has kindly merged this PR and it's scheduled to be included in Gutenberg 14.1, so that's all we need to do here 🎉

**Pull request merged**
